std/range/package.d

```diff
@@ -10761,3 +10764,477 @@ pure unittest
+                else
+                    r.back = val;
+            }
+        }
+
+        static if (isForwardRange!Range)
+        @property typeof(this) save()
+        {
+            return this;
+        }
+
+        static if (isRandomAccessRange!Range)
+        {
+            static if (hasSwappableElements!Range)
+                auto ref opIndex(size_t index) inout
+                {
+                    if (hasFallback) return fallback;
+                    return r[index];
+                }
+            else
+            {
+                auto opIndex(size_t index) inout
+                {
+                    if (hasFallback) return fallback;
+                    return r[index];
+                }
+            }
+        }
+
+        static if (isRandomAccessRange!Range && hasAssignableElements!Range)
+        void opIndexAssign()(auto ref ElementType!Range val, size_t n)
+        {
+            if (hasFallback)
+            {
+                assert(n <= 1, "fallback contains only one element.");
+                fallback = val;
+            }
+            else
+            {
+                r[n] = val;
+            }
+        }
+
+        static if (hasLength!Range)
+        {
+            @property size_t length() const
+            {
+                if (hasFallback) return 1;
+                return r.length;
+            }
+
+            alias opDollar = length;
+        }
+
+        static if (hasSlicing!Range)
+        {
+            auto opSlice(size_t i, size_t j)
+            in
+            {
+                assert(
+                    i <= j,
+                    "Attempting to slice a fallback with a larger first argument than the second."
+                );
+            }
+            body
+            {
+                static if (isInfinite!Range)
+                {
+                    // as an infinite type can't be null, this returns opSlice directly
+                    return r[i .. j];
+                }
+                else
+                {
+                    if (hasFallback)
+                    {
+                        assert(j <= 1, "fallback contains only one element.");
+                        auto ret = typeof(this)(r, fallback);
+                        if (i == 1)
+                            ret.hasFallback = false;
+
+                        return ret;
+                    }
+                    else
+                    {
+                        return typeof(this)(r[i .. j], fallback);
+                    }
+                }
+            }
+
+            static if (isInfinite!Range)
+            {
+                private static struct DollarToken {}
+                enum opDollar = DollarToken.init;
+
+                typeof(this) opSlice(size_t i, DollarToken)
+                {
+                    return typeof(this)(r[i .. $], fallback);
+                }
+            }
+        }
+
+        auto opSlice() { return this; }
+
+        string toString()
```

Comment:
This is odd. Custom toString definitions aren't the norm in phobos because most printing functions recognize ranges